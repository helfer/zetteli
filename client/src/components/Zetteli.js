import React from 'react';

import EditableDateTime from './EditableDateTime';
import EditableTagList from './EditableTagList';
import EditableText from './EditableText';

export default class Zetteli extends React.Component {


    updateText = (evt) => {
       this.props.onUpdate({
           id: this.props.id,
           body: evt.target.value,
       });
    }
    
    onDelete = () => {
        this.props.onDelete(this.props.id);
    }

    render() {
        return (
          <div className="ui card centered fluid">
            <div className="extra content">
              <EditableDateTime datetime={this.props.datetime} /> 
              <span className="right floated">
                <i className="trash icon" onClick={this.onDelete} />
              </span>
              <EditableTagList tags={this.props.tags}/> 
            </div>
            <EditableText text={this.props.body} onChange={this.updateText} /> 
          </div>
        );
    }
}