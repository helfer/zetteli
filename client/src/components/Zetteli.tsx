import * as React from 'react';

import EditableDateTime from './EditableDateTime';
import EditableTagList from './EditableTagList';
import EditableText from './EditableText';

export interface Props {
    id: string;
    body: string;
    tags: string[];
    datetime: Date;
    onUpdate: (arg: {id: string, tags?: string[], body?: string}) => void;
    onDelete: (id: string) => void;
}

export default class Zetteli extends React.Component<Props, object> {
    updateText = evt => {
       this.props.onUpdate({
           id: this.props.id,
           body: evt.target.value,
       });
    }
    
    updateTags = newTags => {
       this.props.onUpdate({
           id: this.props.id,
           tags: newTags,
       });
    }

    onDelete = () => {
        this.props.onDelete(this.props.id);
    }

    render() {
        return (
          <div className="ui card centered fluid">
            <div className="extra content">
              <EditableDateTime datetime={this.props.datetime} /> 
              <span className="right floated">
                <i className="trash icon" onClick={this.onDelete} />
              </span>
              <EditableTagList tags={this.props.tags} updateTags={this.updateTags}/> 
            </div>
            <EditableText text={this.props.body} onChange={this.updateText} /> 
          </div>
        );
    }
}