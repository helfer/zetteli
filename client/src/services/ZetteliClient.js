const LOCAL_STORAGE_KEY = 'Zetteli.zettelis';

export default class ZetteliClient {
    // XXX reading from and writing to local storage
    // will only work in one tab at a time, because
    // writes are not synced and overwrite current contents.
    // That could be solved very easily, but it's not a goal
    // right now.


    constructor() {
        this.zettelis = this.readFromLocalStorage();
    }

    writeToLocalStorage() {
        localStorage.setItem(LOCAL_STORAGE_KEY, JSON.stringify(this.zettelis));
    }

    readFromLocalStorage() {
        return JSON.parse(localStorage.getItem(LOCAL_STORAGE_KEY) || '[]');
    }

    createNewZetteli() {
        const newZetteli = {
            tags: ['log', 'personal'],
            datetime: new Date(),
            body: '',
            id: Math.random(),
        };
        this.zettelis = [ ...this.zettelis, newZetteli ];
        this.writeToLocalStorage();
        return Promise.resolve(true);
    }

    addZetteli(zli) {
        this.zettelis = [ ...this.zettelis, zli];
        this.writeToLocalStorage();
        return Promise.resolve(true);
    }

    removeZetteli(id) {
        this.zettelis = this.zettelis.filter( zli => zli.id !== id );
        this.writeToLocalStorage();
        return Promise.resolve(true);
    }

    updateZetteli(id, data) {
        this.zettelis = this.zettelis.map( zli => {
            if (zli.id === id) {
                return { ...zli, ...data };
            } else {
                return zli;
            }
        });
        this.writeToLocalStorage();
        return Promise.resolve(true);
    }

    getZetteli(id) {
        // TODO: reject when Zetteli cannot be found?
        return Promise.resolve(this.zettelis.find( zli => zli.id === id ));
    }

    getAllZettelis() {
        return new Promise( (resolve, reject) => {
            setTimeout(() => resolve(this.zettelis), 1000);
        });
    }
}