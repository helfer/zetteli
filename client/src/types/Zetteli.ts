export interface ZetteliType {
    id: string,
    body: string,
    tags: string[],
    datetime: Date,
}