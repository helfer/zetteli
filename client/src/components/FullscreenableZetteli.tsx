import Zetteli from './Zetteli';
import Fullscreenable from 'react-fullscreenable';

export default Fullscreenable()(Zetteli);