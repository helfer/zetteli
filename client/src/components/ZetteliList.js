import React from 'react';
import Zetteli from './Zetteli';

export default class ZetteliList extends React.Component {
    state = {
        zettelis: [
            {
                tags: ['log', 'personal'],
                datetime: new Date(),
                body: 'This is Zetteli #1',
            },
            {
                tags: ['note'],
                datetime: new Date(),
                body: 'This is Zetteli #2',
            },
        ],
    }
    render() {
        return (
          <div>
              {this.state.zettelis.map( zli => 
                 <Zetteli
                   tags={zli.tags}
                   datetime={zli.datetime}
                   body={zli.body}
                 />
              )}
              <div>
                <button className="ui circular icon button">
                    <i className="plus icon"></i>
                </button>
              </div>
          </div>
        );
    }
}