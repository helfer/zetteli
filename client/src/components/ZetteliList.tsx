import * as React from 'react';
import Mousetrap from 'mousetrap';

import Zetteli from './Zetteli';
import ZetteliClient from '../services/ZetteliClient';
import { ZetteliType } from './Zetteli';

Mousetrap.prototype.stopCallback = () => false;

export interface Props {
    client: ZetteliClient;
}

export default class ZetteliList extends React.Component<Props, object> {
    state = {
        loading: true,
        zettelis: [] as ZetteliType[],
    };

    refetchZettelis() {
        // TODO: Because this will only work if the call doesn't take too long.
        this.props.client.getAllZettelis().then( zettelis => {
            this.setState({ zettelis });
        });
    }

    createNewZetteli = () => {
        this.props.client.createNewZetteli()
        .then(() => this.refetchZettelis());
    }

    updateZetteli = (modifiedZetteli: ZetteliType) => {
        this.props.client.updateZetteli(modifiedZetteli.id, modifiedZetteli)
        .then(() => this.refetchZettelis());
    }

    deleteZetteli = (id: string) => {
        this.props.client.deleteZetteli(id)
        .then(() => this.refetchZettelis());
    }

    componentWillMount() {
        this.props.client.getAllZettelis().then( zettelis => {
            this.setState({ zettelis, loading: false });
        });
    }

    componentDidMount() {
        Mousetrap.bind(['command+u'], this.createNewZetteli);

    }
    componentWillUnmount() {
        Mousetrap.unbind(['command+u']);
    }

    render() {
        if (this.state.loading) {
            return (
                <div className="ui active inverted dimmer">
                    <div className="ui text loader">Loading</div>
                </div>
            );
        }
        return (
          <div style={{ marginBottom: '20em' }} className="zetteliList">
              {this.state.zettelis.map( zli => 
                 <Zetteli
                   tags={zli.tags}
                   datetime={zli.datetime}
                   body={zli.body}
                   key={zli.id}
                   id={zli.id}
                   onUpdate={this.updateZetteli}
                   onDelete={this.deleteZetteli}
                 />
              )}
              <div style={{textAlign: 'center'}}>
                <button className="ui center aligned circular icon button" onClick={this.createNewZetteli}>
                    <i className="plus icon" />
                </button>
            </div>
          </div>
        );
    }
}