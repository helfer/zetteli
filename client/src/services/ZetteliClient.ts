import { ZetteliType } from '../components/Zetteli';

export interface ZetteliClient {
    createNewZetteli(): Promise<boolean>;
    addZetteli(zli: ZetteliType): Promise<boolean>;
    deleteZetteli(id: string): Promise<boolean>;
    updateZetteli(id: string, data: ZetteliType): Promise<boolean>;
    getZetteli(id: string): Promise<ZetteliType | undefined>;
    getAllZettelis(): Promise<ZetteliType[]>;
}