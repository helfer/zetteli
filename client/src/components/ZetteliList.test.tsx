import * as React from 'react';
import * as enzyme from 'enzyme';
import FileSaver from 'file-saver';
import Mousetrap from 'mousetrap';

import ZetteliList from './ZetteliList';
import { Props as ZetteliProps, ZetteliType } from './Zetteli';
import FullscreenableZetteli from './FullscreenableZetteli';
import ZetteliClient from '../services/ZetteliClient';

describe('ZetteliList', () => {
    let client: ZetteliClient;
    let store: Storage;
    let storage: Map<string, string>;
    let zli: ZetteliType;
    let zli2: ZetteliType;

    beforeEach(() => {
        // TODO(helfer): Make a proper mock of ZetteliClient
        storage = new Map();
        store = {
            getItem: key => storage.get(key) || null,
            length: storage.size,
            clear: () => { storage.clear(); },
            setItem: (key, item) => storage.set(key, item),
            removeItem: key => storage.delete(key),
            key: i => '', // not implemented
        };

        client = new ZetteliClient(store);
        zli = {
            id: '1',
            body: 'hello',
            datetime: new Date(1500000000000),
            tags: ['t1', 't2'],
        };
        client.addZetteli(zli);
        zli2 = {
            id: '2',
            body: 'hello2',
            datetime: new Date(1500000005000),
            tags: ['t3', 't4'],
        };
        client.addZetteli(zli2);
    });

    it('renders loading state while waiting', () => {
        const zetteliList = enzyme.shallow(<ZetteliList client={client} />);
        expect(zetteliList.text()).toContain('Loading');
    });
    it('renders the list of zettelis with correct props', () => {
        const zetteliList = enzyme.shallow(<ZetteliList client={client} />);
        // NOTE(helfer): It would be better to not be aware of state, but I can't
        // figure out how to successfuly let the ZetteliClient promise update the state.
        zetteliList.setState({ zettelis: [zli, zli2], loading: false });
        zetteliList.update();
        // TODO(helfer): Fix the type declaration of the Fullscreenable HOC soyou don't need to coerce type
        expect((zetteliList.find(FullscreenableZetteli).at(0).props() as ZetteliProps).id).toBe('1');
        expect((zetteliList.find(FullscreenableZetteli).at(0).props() as ZetteliProps).onUpdate).toBeDefined();
        expect((zetteliList.find(FullscreenableZetteli).at(1).props() as ZetteliProps).id).toBe('2');
        expect((zetteliList.find(FullscreenableZetteli).at(1).props() as ZetteliProps).onDelete).toBeDefined();
    });
    // listens to cmd+u and creates a new Zetteli if it's pressed
    // TODO(helfer): that test is actually a bit tricky. Let's keep that for later
    it('listens to command+u and adds a new zetteli if called', () => {
        let cb: Function = () => null;
        Mousetrap.bind = jest.fn( (key, func) => {
            // avoid capturing ctrl+s / meta+s
            if (key === 'command+u') {
                cb = func;
            }
        });
        Mousetrap.unbind = jest.fn();
        client.createNewZetteli = jest.fn(() => Promise.resolve([]));

        // NOTE(helfer): Have to mount here to get lifecycle events
        const zetteliList = enzyme.mount(<ZetteliList client={client} />);
        expect(Mousetrap.bind).toHaveBeenCalled();
        cb(); // Should have been assigned by now
        expect(client.createNewZetteli).toHaveBeenCalled();
        zetteliList.unmount();
        expect(Mousetrap.unbind).toHaveBeenCalledWith('command+u');

        // TODO(helfer): Do the mock functions get removed after every test?
    });

    it('downloadZettelis saves zettelis to file', () => {
        const mockSaveAs = jest.fn();
        FileSaver.saveAs = mockSaveAs;
        // TODO(helfer): Find out why global.Blob is not defined
        // tslint:disable-next-line no-any
        (global as any).Blob = (content: string, options: {}) => ({content, options}); 
        const zetteliList = enzyme.shallow(<ZetteliList client={client} />);
        zetteliList.setState({ zettelis: [zli, zli2], loading: false });
        zetteliList.update();
        (zetteliList.instance() as ZetteliList).downloadZettelis();
        expect(mockSaveAs.mock.calls[0][0]).toMatchSnapshot();
    });

    // deleteZetteli calls client.delete
    it('can delete a zetteli', () => {
        client.deleteZetteli = jest.fn(() => Promise.resolve());
        const zetteliList = enzyme.shallow(<ZetteliList client={client} />);
        // TODO(helfer): Why can't I just call the prop that was passed?
        (zetteliList.instance() as ZetteliList).deleteZetteli('2');
        expect(client.deleteZetteli).toHaveBeenCalledWith('2');
    });
    // createNewZetteli calls client.create
    it('can create a zetteli', () => {
        client.createNewZetteli = jest.fn(() => Promise.resolve());
        const zetteliList = enzyme.shallow(<ZetteliList client={client} />);
        (zetteliList.instance() as ZetteliList).createNewZetteli();
        expect(client.createNewZetteli).toHaveBeenCalled();
    });
    // updateZetteli calls client.update
    it('can update a zetteli', () => {
        client.updateZetteli = jest.fn(() => Promise.resolve());
        const zetteliList = enzyme.shallow(<ZetteliList client={client} />);
        (zetteliList.instance() as ZetteliList).updateZetteli({ ...zli2, id: zli.id });
        expect(client.updateZetteli).toHaveBeenCalledWith(zli.id, { ...zli2, id: zli.id });
    });
    // refetch calls getAllZettelis
    it('can refetch zettelis', () => {
        client.getAllZettelis = jest.fn(() => Promise.resolve([]));
        const zetteliList = enzyme.shallow(<ZetteliList client={client} />);
        (zetteliList.instance() as ZetteliList).refetchZettelis();
        expect(client.getAllZettelis).toHaveBeenCalled();
    });
});