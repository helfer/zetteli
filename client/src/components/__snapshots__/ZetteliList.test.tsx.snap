// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ZetteliList downloadZettelis saves zettelis to file 1`] = `
Object {
  "content": Array [
    "
---
>2017-07-06T19:40:00-07:00
[t1 t2]

hello

            ",
    "
---
>2017-07-06T19:40:05-07:00
[t3 t4]

hello2

            ",
  ],
  "options": Object {
    "type": "text/plain;charset=utf-8",
  },
}
`;
